### PR TITLE
[MM-17554] Health Check: improves enhanced health check with a DB write check.

### PR DESCRIPTION
**Pull request by jesperhansen17**
#### Summary
This PR adds a DB write check to the enhanced ping health check function. It writes a timestamp to `System` table and also performs a read check by reading `System` table and compare the timestamp saved in the DB.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/11813

### CODE REVIEW

File: api4/system.go [hunk 1]
```diff
@@ -64,11 +65,32 @@ func getSystemPing(c *Context, w http.ResponseWriter, r *http.Request) {
...
 			s[dbStatusKey] = model.STATUS_UNHEALTHY
 			s[model.STATUS] = model.STATUS_UNHEALTHY
+		} else {
+			healthCheck, readErr := c.App.Srv.Store.System().GetByName(healthCheckKey)
+			if readErr != nil {
+				mlog.Debug(fmt.Sprintf("unable to read from database: %s", readErr.Error()))
+				s[dbStatusKey] = model.STATUS_UNHEALTHY
+				s[model.STATUS] = model.STATUS_UNHEALTHY
+			} else if healthCheck.Value != currentTime {
+				mlog.Debug(fmt.Sprintf("incorrect healthcheck value, expected %s, got %s", healthCheck.Value, currentTime))
```
Comment (gabrieljackson) [inline api4/system.go]: It looks like we need to switch around the expected + actual values here.

```suggestion
				mlog.Debug(fmt.Sprintf("Incorrect healthcheck value: expected %s, got %s", currentTime, healthCheck.Value))
```

File: api4/system.go [hunk 2]
```diff
@@ -64,11 +65,32 @@ func getSystemPing(c *Context, w http.ResponseWriter, r *http.Request) {
...
+		// Database Write/Read Check
+		currentTime := fmt.Sprintf("%d", time.Now().Unix())
+		healthCheckKey := "health_check"
+
+		writeErr := c.App.Srv.Store.System().SaveOrUpdate(&model.System{
+			Name:  healthCheckKey,
+			Value: currentTime,
+		})
+		if writeErr != nil {
+			mlog.Debug(fmt.Sprintf("unable to write to database: %s", writeErr.Error()))
```
Comment (gabrieljackson) [inline api4/system.go]: nit: It looks like our other ping debug logs start with capital letters so doing the same here would be ideal.

File: api4/system.go (hunk 1 above)
Comment (jesperhansen17) [reply api4/system.go]: Yes, good catch.

File: api4/system.go (hunk 2 above)
Comment (jesperhansen17) [reply api4/system.go]: True, I will update all the logging to be uppercase.

**Comment by jesperhansen17** [comment]
@gabrieljackson Regarding the unit tests, I had a look at them and as far as I can tell there was no unit test covering the DB part.

I looked into how to create unit tests to cover DB write/read error, but I couldn't figure out how to mock the `store` responses to be off type `appErr`.

So I opened this PR so I could hopefully receive some help or some can point me in a directions where this has been done before?

**Comment by srkgupta** [comment]
This change will be tested as part of the 5.16 release. Removing the QA Review label.

**Comment by gabrieljackson** [comment]
@jesperhansen17, that's a good question! We don't seem to have a great way to mock out the exact store calls we are using, but by using `th.App.UpdateConfig()` to change the DB connection string to something bad, we can simulate a complete lack of connection to the DB which should allow us to make a fairly good unit test.

Information on the config settings can be found here: https://docs.mattermost.com/administration/config-settings.html#database

**Comment by gabrieljackson** [comment]
@jesperhansen17, did you have any further questions on this? I am happy to help with anything else you may need.

**Comment by jesperhansen17** [comment]
Yeah @gabrieljackson, been unavailable for a couple of days.

But I do have a question, I'm struggling to simulate the bad DB connection by updating the connection string by using `UpdateConfig` you referring to.

```
th.App.UpdateConfig(func(cfg *model.Config) {
    cfg.SqlSettings.DataSource = &invalidConnectionString
})
```

I'm must be missing something because I'm not receiving the output wanted. The docs above states that the server needs to be restarted before changes taking effect. 

So currently I'm quite stuck getting this unit tests to run.

**Comment by gabrieljackson** [comment]
@jesperhansen17, after reviewing our test interfaces for the `api4` package, I have come to the same conclusion you have. There doesn't seem to be an elegant way to put the database in the state we want and, at the same time, updating the database connection info requires a restart.

We will want tests for the functionality introduced in this PR, but that can happen in a follow-up change as that work is beyond the scope of this ticket.

If you can merge the master branch into this branch and push up the commit, we can get tests green and will merge it asap. Thanks for your patience!

**Comment by jesperhansen17** [comment]
@gabrieljackson, no problem.

I have merged master branch and everything seems fine.

Thanks for all the help :)

**Pull request merged**